File: packages/csharp/SourceRpc.SocketIo/SocketIoClientTransport.cs
using System.Text;
using System.Text.Json;
using MessagePack;
using MessagePack.Resolvers;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using SocketIOClient;

namespace SourceRpc.SocketIo;

/// <summary>
/// How a flat frame is encoded before it becomes a socket.io event payload.
///
/// The choice is not socket.io's - socket.io carries whatever it is given and would happily carry
/// the frame as a plain object. It exists because the TypeScript side runs a codec over the event
/// payload either way, so what travels is always bytes and the codec decides what is inside them.
/// Both ends must have made the same choice, and neither can tell from the wire which the other
/// made: an unreadable frame is reported as one, not guessed at.
/// </summary>
public enum SocketIoFrameCodec
{
    /// <summary>MessagePack. The TypeScript library's default, so this is the one to leave alone.</summary>
    MessagePack,

    /// <summary>UTF-8 JSON, for a server configured with a JSON codec, or for reading a capture.</summary>
    Json
}

/// <summary>Where a socket.io peer connects, and what it says on the way in.</summary>
public sealed class SocketIoTransportOptions
{
    /// <summary>The server, e.g. <c>http://localhost:3000</c>.</summary>
    public string Url { get; set; } = "http://localhost:3000";

    /// <summary>
    /// Credentials for the handshake, which is where a socket.io server authenticates.
    ///
    /// This is the whole of identity on this carrier and the reason frames here are not signed: the
    /// connection is authenticated once, and the server pins every frame's `src` to the identity it
    /// authenticated. A per-frame signature would be a weaker claim checked in more places. MQTT
    /// cannot do this - a broker gives a receiver no connection to attribute a message to - which is
    /// why SourceRpc.Mqtt has signing and this does not.
    /// </summary>
    public object? Auth { get; set; }

    /// <inheritdoc cref="SocketIoFrameCodec"/>
    public SocketIoFrameCodec Codec { get; set; } = SocketIoFrameCodec.MessagePack;

    /// <summary>
    /// Announce this peer on connect. Off leaves it unlisted, and so unaddressable by anyone who
    /// looks the network up rather than being told a name.
    /// </summary>
    public bool AnnouncePresence { get; set; } = true;

    /// <summary>The namespace path, for a server that serves more than one.</summary>
    public string? Path { get; set; }

    /// <summary>Applied to the options socket.io itself is built with, for anything not exposed here.</summary>
    public Action<SocketIOOptions>? Configure { get; set; }
}

/// <summary>
/// A link to a socket.io server, for a .NET peer that dials out.
///
/// The third binding, and the one that shows the seam was worth cutting: it shares its frame layout
/// with SignalR, its presence handling with both, and its retry shape with all three, so what is
/// actually written here is an encoder, a decoder and two event names. Correlation, deadlines,
/// tickets, fences and error mapping are <see cref="SourceRpcClient"/>'s and
/// <see cref="RpcDispatcher"/>'s, and stay identical across every carrier.
///
/// **Client only.** socket.io's server is a Node library with no maintained .NET equivalent, so a
/// .NET process that needs to be dialled into should serve SignalR instead - which is the same flat
/// frame under different method names, and is why a TypeScript client can reach either.
///
/// Mirrors the TypeScript <c>SocketIoClientTransport</c>: <c>frame</c> carries codec-encoded bytes,
/// <c>presence</c> carries an ordinary object, and both are announced on every connect rather than
/// only the first - a server forgets a peer when its socket drops, so a reconnected peer that stayed
/// quiet would be unreachable while appearing perfectly healthy.
/// </summary>
public sealed class SocketIoClientTransport : ISourceRpcTransport
{
    private readonly SourceRpcOptions _options;
    private readonly SocketIoTransportOptions _socketIo;
    private readonly ILogger _log;
    private SocketIO? _socket;
    private CancellationTokenSource? _closing;

    /// <inheritdoc/>
    public string Name => _options.Name;

    /// <inheritdoc/>
    public bool Connected => _socket?.Connected == true;

    /// <inheritdoc/>
    public event Func<RpcFrame, Task>? FrameReceived;

    /// <inheritdoc/>
    public event Action<IReadOnlyCollection<string>>? PeersChanged;

    /// <summary>Connect to a socket.io server at a URL.</summary>
    public SocketIoClientTransport(string url, SourceRpcOptions options, ILogger? log = null)
        : this(new SocketIoTransportOptions { Url = url }, options, log)
    {
    }

    /// <summary>Connect with the full option set.</summary>
    public SocketIoClientTransport(SocketIoTransportOptions socketIo, SourceRpcOptions options, ILogger? log = null)
    {
        _socketIo = socketIo;
        _options = options;
        _log = log ?? NullLogger.Instance;
    }

    /// <inheritdoc/>
    public Task StartAsync(CancellationToken cancellationToken = default)
    {
        if (_socket is not null)
            return Task.CompletedTask;
        _closing = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

        var options = new SocketIOOptions
        {
            // socket.io's own reconnection, kept rather than replaced: unlike SignalR's it does
            // cover a failed first attempt, so a peer that starts before its server is simply not
            // connected yet. The loop below exists only for what it does not cover - a server that
            // refuses the handshake outright, which stops socket.io retrying for good.
            Reconnection = true,
            ReconnectionDelayMax = _options.ReconnectDelaysMs.Length > 0 ? _options.ReconnectDelaysMs[^1] : 5000,
            ReconnectionAttempts = int.MaxValue,
            Auth = _socketIo.Auth
        };
        if (_socketIo.Path is { Length: > 0 } path)
            options.Path = path;
        _socketIo.Configure?.Invoke(options);

        var socket = new SocketIO(new Uri(_socketIo.Url), options);
        _socket = socket;

        socket.On(TransportContract.FrameName, OnFrameAsync);
        socket.On(TransportContract.PresenceName, OnPresenceAsync);
        // EventHandler, so the announcement cannot be awaited here and is fired instead. Failing to
        // announce is logged rather than thrown: the link is up either way, and a peer that cannot
        // be found is a better state to be in than one that took the process down.
        socket.OnConnected += (_, _) => _ = AnnounceAsync();
        socket.OnDisconnected += (_, reason) =>
        {
            // Nothing is reachable through a link that is down, and the server sends a fresh
            // snapshot when it returns. Reported empty so a console can grey the whole list out
            // rather than showing peers that cannot be called.
            _log.LogWarning("SourceRpc socket.io link closed: {Reason}", reason);
            PeersChanged?.Invoke([]);

            // socket.io deliberately never auto-reconnects after a *server-initiated* close - and
            // that is exactly what a restarting server sends on its way down. Left alone, one plant
            // server reboot orphans every .NET peer it had, permanently and silently: the process
            // is healthy, sends throw "not connected", and nothing ever tries again.
            //
            // Retried on a delay, so a server that closes deliberately - a displacement, a refused
            // handshake - is not hammered. Two peers configured with one name will fight through
            // this; that is the misconfiguration's noise, not a reason to stay dead after every
            // reboot. The TypeScript client transport carries the same workaround, for the reason.
            if (reason == ServerInitiatedClose && _closing?.IsCancellationRequested == false)
                _ = RetryAfterServerCloseAsync();
        };

        // Not awaited: ConnectAsync does not return until it is connected, and a peer that starts
        // before its server would otherwise block its own startup on something that may be minutes
        // away. A send meanwhile throws rather than being discarded, which tells a caller now.
        _ = ConnectAsync();
        return Task.CompletedTask;
    }

    /// <summary>
    /// What socket.io calls a close the server asked for, rather than a link that failed.
    /// </summary>
    private const string ServerInitiatedClose = "io server disconnect";

    /// <summary>Come back after a server closed the link, which socket.io will not do on its own.</summary>
    private async Task RetryAfterServerCloseAsync()
    {
        try
        {
            await Task.Delay(1000, _closing!.Token);
            if (_closing?.IsCancellationRequested == false && _socket?.Connected == false)
                await ConnectAsync();
        }
        catch (OperationCanceledException)
        {
            // Closed while waiting, which is the ordinary way this ends.
        }
    }

    /// <summary>
    /// Connect, and keep trying.
    ///
    /// Never throws. socket.io retries a dropped link on its own, but gives up permanently when a
    /// server refuses the handshake - and a plant server refuses handshakes for the whole of its
    /// reboot. Left alone, that orphans every client it had.
    /// </summary>
    private async Task ConnectAsync(int attempt = 0)
    {
        var socket = _socket;
        if (socket is null || _closing?.IsCancellationRequested != false)
            return;
        try
        {
            await socket.ConnectAsync();
            _log.LogInformation("SourceRpc connected to socket.io as {Peer}", _options.Name);
        }
        catch (Exception e) when (_closing?.IsCancellationRequested == false)
        {
            var delay = _options.ReconnectDelaysMs[Math.Min(attempt, _options.ReconnectDelaysMs.Length - 1)];
            _log.LogWarning(e, "SourceRpc could not reach the socket.io server; retrying in {Delay}ms", delay);
            try
            {
                await Task.Delay(delay, _closing!.Token);
                await ConnectAsync(attempt + 1);
            }
            catch (OperationCanceledException)
            {
                // Closed while waiting, which is the ordinary way this loop ends.
            }
        }
    }

    /// <summary>
    /// Say who this peer is.
    ///
    /// On every connect, not only the first: the server keeps its peer table per socket, so a
    /// reconnected peer that stayed silent is one nobody can address - and the failure looks like a
    /// working link, because frames still leave.
    /// </summary>
    private async Task AnnounceAsync()
    {
        if (!_socketIo.AnnouncePresence)
            return;
        var socket = _socket;
        if (socket is null)
            return;
        try
        {
            // As an object rather than as bytes: presence is the one message the TypeScript side
            // does not run its codec over, on either end.
            await socket.EmitAsync(
                TransportContract.PresenceName,
                [new PresenceAnnouncement { Name = _options.Name, V = TransportContract.FrameVersion }]);
        }
        catch (Exception e)
        {
            _log.LogWarning(e, "SourceRpc could not announce {Peer}", _options.Name);
        }
    }

    /// <summary>
    /// Read one frame off the socket and start handling it - without waiting for that to finish.
    ///
    /// The decode happens here, synchronously, because the context belongs to the read loop and is
    /// not valid once this returns. The **dispatch** deliberately does not: socket.io awaits this
    /// callback before reading the next message, so awaiting a responder here means nothing else
    /// arrives until it returns - and a responder that calls out while handling a call would then
    /// wait for a reply that cannot be read until it stops waiting. That is a deadlock, and it
    /// resolves only when the outer call times out, which reads as a slow peer rather than as this.
    ///
    /// The TypeScript client does the same thing for the same reason: socket.io's JavaScript client
    /// never awaits a listener either, so this is the ordering the protocol was written against
    /// rather than a liberty taken here. Frames are correlated and events carry a cursor, so nothing
    /// above depends on them being *handled* in arrival order.
    /// </summary>
    private Task OnFrameAsync(IEventContext context)
    {
        RpcFrame frame;
        try
        {
            // A frame arrives as a binary attachment, so anything else on this event is not one.
            frame = Decode(context.GetValue<byte[]>(0) ?? throw new FormatException("frame event carried no payload"));
        }
        catch (Exception e)
        {
            // Announced, never dropped. Silence reaches a caller as a timeout, which is
            // indistinguishable from a slow method and sends the search to the wrong place.
            _log.LogError(e, "SourceRpc could not read an inbound socket.io frame");
            return Task.CompletedTask;
        }

        _ = DispatchAsync(frame);
        return Task.CompletedTask;
    }

    private async Task DispatchAsync(RpcFrame frame)
    {
        var handler = FrameReceived;
        if (handler is null)
            return;
        try
        {
            await handler(frame);
        }
        catch (Exception e)
        {
            // One bad frame from the far end must not take this peer down.
            _log.LogError(e, "SourceRpc failed to handle an inbound frame");
        }
    }

    private Task OnPresenceAsync(IEventContext context)
    {
        try
        {
            var update = context.GetValue<PresenceUpdate>(0);
            if (update?.Peers is { Length: > 0 } peers)
                PeersChanged?.Invoke(peers);
        }
        catch (Exception e)
        {
            _log.LogWarning(e, "SourceRpc could not read a presence update");
        }
        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    public async Task SendAsync(RpcFrame frame, CancellationToken cancellationToken = default)
    {
        var socket = _socket;
        if (socket is null || !socket.Connected)
            // Thrown rather than dropped: a frame discarded in silence leaves its caller waiting out
            // the whole deadline for something that was never going to be sent.
            throw new SourceRpcException(RpcErrorCode.TransportError, "not connected to the socket.io server");
        await socket.EmitAsync(TransportContract.FrameName, [Encode(frame)], cancellationToken);
    }

    /// <summary>
    /// The frame as the bytes a socket.io event carries.
    ///
    /// Bytes rather than an object because the far end decodes the payload with a codec either way,
    /// so an object would arrive as something its decoder cannot read.
    /// </summary>
    private byte[] Encode(RpcFrame frame) =>
        _socketIo.Codec == SocketIoFrameCodec.Json
            ? JsonSerializer.SerializeToUtf8Bytes(frame, JsonOptions)
            // Contractless, so a body that is a record or an anonymous type writes as a map instead
            // of failing in the primitive formatter - the same resolver the MQTT binding uses, for
            // the same reason.
            : MessagePackSerializer.Serialize(frame, ContractlessStandardResolver.Options);

    private RpcFrame Decode(byte[] payload) =>
        _socketIo.Codec == SocketIoFrameCodec.Json
            ? JsonSerializer.Deserialize<RpcFrame>(payload, JsonOptions) ?? throw new FormatException("empty JSON frame")
            : MessagePackSerializer.Deserialize<RpcFrame>(payload, ContractlessStandardResolver.Options);

    /// <summary>
    /// The frame's own names are pinned by attribute, so nothing here renames anything. What this
    /// settles is the <c>body</c>, which is <see cref="object"/> and would otherwise decode to a
    /// <see cref="JsonElement"/> under one reader and something else under another.
    /// </summary>
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        // Idempotent, because a transport is ordinarily disposed twice: SourceRpcClient owns it and
        // disposes it, and a caller's `await using` on the transport does so again.
        var closing = Interlocked.Exchange(ref _closing, null);
        var socket = Interlocked.Exchange(ref _socket, null);
        if (closing is not null)
        {
            await closing.CancelAsync();
            closing.Dispose();
        }
        if (socket is not null)
        {
            try
            {
                await socket.DisconnectAsync();
            }
            catch (Exception e)
            {
                // A link already down cannot be closed again, and that is not a failure to report.
                _log.LogDebug(e, "SourceRpc socket.io disconnect failed on dispose");
            }
            socket.Dispose();
        }
    }
}

File: packages/rpc/src/Transports/SocketIoClientTransport.ts
import { io, ManagerOptions, Socket, SocketOptions } from 'socket.io-client'
import { GenericModule, IGenericModule, Message, publicTransportEndpoint, TransportEvent, type RpcTransportDescription } from '../RPC/Core.js'
import { FrameCodec, msgPackCodec } from '../RPC/Codec.js'
import { refuseDelivery } from '../RPC/Undeliverable.js'
import { isUsablePeerName, isUsableShape, MAX_RELAY_HOPS, PRESENCE_EVENT, PresenceAnnouncement, PresenceUpdate } from './Presence.js'
import { FRAME_EVENT, fromWireFrame, LEGACY_FRAME_EVENT, FLAT_FRAME_VERSION, toWireFrame } from './FlatFrame.js'

export class SocketIoClientTransport extends GenericModule<Message, unknown, Message, unknown> {
    socket?: Socket
    connected = false
    /** Owned here rather than by a converter above, so the transport decides its own wire form. */
    codec: FrameCodec = msgPackCodec
    /**
     * Send the flat frame rather than the `$`-delimited one. See docs/flat-frame-spec.md.
     *
     * Both are read on the way in, so a server of either vintage can answer this peer. What this
     * decides is what goes *out*, and the honest limit is worth stating: a v2 frame reaches a
     * pre-v2 server as an event it has no listener for, which socket.io delivers to nobody - so the
     * call times out with nothing said. That is why this is a flag and not merely a rewrite, and
     * why `rpc` and its dependants version together.
     */
    frameVersion: 1 | 2 = FLAT_FRAME_VERSION
    /** Peers this transport has been told are online, so a reconnect can report only what changed. */
    readonly knownPeers = new Set<string>()
    /** Peers reachable through whatever owns this transport, advertised to the far end. */
    private carrying: string[] = []
    /** Resolves the sweep below. Undefined once it has fired - settled never unsettles. */
    private sweepLanded?: () => void
    private readonly sweep = new Promise<void>((resolve) => (this.sweepLanded = resolve))

    /**
     * Resolved when the server's answer to this transport's announcement - the full peers list -
     * has arrived. That list is socket.io's stand-in for MQTT's retained presence, so this is the
     * moment "who is here?" stops being answered from an empty registry that merely has not been
     * filled yet. Resolved once and stays resolved: settled is about the first picture, not about
     * every change since, and a reconnect's fresh snapshot does not re-open it.
     *
     * A transport that does not announce is settled immediately, because the server only sends the
     * list in reply to an announcement - nothing is coming, and waiting for it would just spend
     * the caller's bound.
     */
    presenceSettled(): Promise<void> {
        if (!this.announcePresence) this.sweepLanded?.()
        return this.sweep
    }

    constructor(
        /**
         * The peer name this transport announces itself under, which is what makes it findable and
         * addressable. Taking it here mirrors MqttTransport, where the name has always been needed
         * to subscribe to the peer's own topic.
         */
        name: string,
        public url?: string,
        sources?: IGenericModule[],
        // SocketOptions carries `auth`, which is how credentials reach an authenticating server.
        public options: Partial<ManagerOptions & SocketOptions> = {},
        /** Announce on connect. Off leaves this peer unlisted and unaddressable, as before. */
        public announcePresence = true,
        /**
         * Connect to an `https://` or `wss://` server without checking its certificate.
         *
         * This used to be the default, and it should never have been: it accepts any certificate at
         * all, so anything able to answer on the server's address can read and rewrite everything
         * sent over the link - which on this library's traffic means industrial commands. Left as a
         * deliberate, named choice for a development server with a self-signed certificate. A plant
         * with its own certificate authority should pass `ca` in the socket options instead, which
         * keeps verification on.
         */
        public allowInsecureTls = false,
        /** Which frame layout to send. See the `frameVersion` field for what the older one costs. */
        frameVersion: 1 | 2 = FLAT_FRAME_VERSION
    ) {
        super(name, sources)
        this.frameVersion = frameVersion
        // Deferred by a microtask so whatever constructs this transport can finish wiring it
        // before the link comes up. A resumed MQTT session is delivered its queued messages the
        // instant it connects, and a frame arriving before the RPC handler is piped in would find
        // no target and be dropped. A fresh session never exposes this, because nothing arrives
        // that early.
        queueMicrotask(() => void this.open().catch((e) => this.emit(TransportEvent.transportError, e)))
    }

    rpcDescription(): RpcTransportDescription {
        const endpoint = this.url ? publicTransportEndpoint(this.url) : undefined
        return { name: this.getName(), protocol: 'socket.io', role: 'connect', ...(endpoint ? { endpoint } : {}) }
    }

    /** Set for good by close(). What stops the server-restart retry from resurrecting the link. */
    private closing = false

    override async close() {
        this.closing = true
        // A waiter on the sweep must not outlive the transport - on a link that never came up,
        // the answer to "has the first picture arrived" is that no picture is coming.
        this.sweepLanded?.()
        this.sweepLanded = undefined
        const socket = this.socket
        this.socket = undefined
        this.connected = false
        this.readyFlag = false
        if (!socket) return
        // Disarm reconnection before disconnecting. An explicit close is not a link failure, and
        // a manager left free to reconnect keeps a timer armed that outlives the transport.
        // reconnection(false) is the setter; assigning to opts.reconnection does not reach the
        // manager's own flag and left the timer armed anyway.
        socket.io.reconnection(false)
        // Only disconnect() - close() is an alias for it, and calling both corrupted the manager's
        // socket bookkeeping.
        //
        // Awaited, because disconnect() only *starts* the close: it sends a close packet and
        // returns, leaving the engine's ping timer armed until the transport is actually torn down.
        // Returning before that makes close() a promise that resolves while the connection it was
        // supposed to close is still running.
        const engine = socket.io.engine
        const closed =
            socket.connected && engine
                ? new Promise<void>((resolve) => {
                      // Bounded: a close that never completes must not hang the caller forever.
                      const settle = setTimeout(resolve, 2000)
                      settle.unref?.()
                      engine.once('close', () => {
                          clearTimeout(settle)
                          resolve()
                      })
                  })
                : Promise.resolve()
        socket.disconnect()
        await closed
        socket.removeAllListeners()
        this.knownPeers.clear()
    }

    override async open() {
        // Idempotent: the constructor opens, and RpcClient.init() opens again. Without this guard
        // every client ends up with a second, orphaned socket that stays connected forever.
        if (this.socket) return
        // Deliberately not awaited. The base hook is a no-op, and awaiting it yields before the
        // socket below is assigned - which lets a second open() past the guard above and leaves the
        // first socket orphaned and reconnecting forever, exactly what the guard is here to prevent.
        void super.open()
        const urlSocketIo = this.url
        // Certificate verification is Node's default and stays on. It was turned off here for
        // every client, before the caller's own options were applied, so a Node peer accepted an
        // impersonated TLS server unless whoever wrote it knew to turn verification back on.
        if (this.allowInsecureTls) {
            // Spread after, so a caller that asks for both gets the safer of the two.
            this.options = { rejectUnauthorized: false, ...this.options }
            this.warnAboutInsecureTls()
        }
        this.socket = urlSocketIo ? io(urlSocketIo, this.options) : io(this.options)
        this.socket.on(LEGACY_FRAME_EVENT, async (messageArray) => {
            try {
                const [header, payload, reason] = this.extractHeader(new Uint8Array(messageArray))
                if (!header) {
                    // Reported rather than dropped in silence, which showed up only as a timeout.
                    this.emit(TransportEvent.rejected, { source: 'unknown', reason: reason ?? 'no msgrpc header' })
                    return
                }
                const message = this.codec.decode(payload as Uint8Array) as Message
                await this.deliver(message, header.source, header.target, header.hops ?? 0)
            } catch (e) {
                // A peer that sends a frame this codec cannot read must not take the client down.
                this.emit(TransportEvent.rejected, { source: 'unknown', reason: `undecodable frame: ${String(e)}`, error: e })
            }
        })
        // Both layouts are read regardless of which one this peer sends, so a server may answer in
        // whichever it speaks and an upgrade needs no coordination in this direction.
        this.socket.on(FRAME_EVENT, async (frameArray) => {
            try {
                const read = fromWireFrame(this.codec.decode(new Uint8Array(frameArray)))
                if ('reason' in read) {
                    this.emit(TransportEvent.rejected, { source: 'unknown', reason: read.reason })
                    return
                }
                await this.deliver(read.message, read.source, read.target, read.hops)
            } catch (e) {
                this.emit(TransportEvent.rejected, { source: 'unknown', reason: `undecodable frame: ${String(e)}`, error: e })
            }
        })
        this.socket.on(PRESENCE_EVENT, (update: PresenceUpdate) => {
            // socket.io emits synchronously from its parser, so a listener that throws unwinds into
            // the engine rather than anywhere that could report it.
            try {
                this.onPresence(update)
            } catch (e) {
                this.emit(TransportEvent.rejected, { source: 'unknown', reason: `bad presence update: ${String(e)}`, error: e })
            }
        })
        // socket.io emits 'connect' on reconnects too, so this fires on every transition.
        this.socket.on('connect', () => {
            this.connected = true
            this.readyFlag = true
            // Announced on every connect, not only the first: the server forgets a peer when its
            // socket drops, so a reconnected peer that stayed silent would be unaddressable.
            if (this.announcePresence) this.announce()
            this.emit(TransportEvent.connected)
        })
        this.socket.on('disconnect', (reason) => {
            this.connected = false
            this.readyFlag = false
            // Nothing is reachable through a link that is down, and the server will send a fresh
            // snapshot when it comes back. Reported so a console can grey the whole list out.
            for (const peer of [...this.knownPeers]) {
                this.knownPeers.delete(peer)
                this.emit(TransportEvent.peerGone, peer)
            }
            this.emit(TransportEvent.disconnected, reason)
            // socket.io deliberately never auto-reconnects after 'io server disconnect' - and that
            // is exactly what a restarting server sends on its way down. Left alone, a plant
            // server reboot permanently orphans every client it had, which no HMI can accept. So a
            // server-initiated close is retried like a network drop, on a delay so a server that
            // kicks deliberately - displacement, a rejected handshake - is not hammered. Two peers
            // configured with one name will fight through this; that is the misconfiguration's
            // noise, not a reason to stay disconnected after every reboot.
            if (reason === 'io server disconnect' && !this.closing) {
                const retry = setTimeout(() => {
                    if (!this.closing && this.socket?.disconnected) this.socket.connect()
                }, 1000)
                retry.unref?.()
            }
        })
    }

    /**
     * Said once per transport, and only where it means anything: the flag has no effect on a plain
     * `http://` link, and warning about one would teach people to ignore the warning.
     */
    private warnAboutInsecureTls() {
        if (!this.url || !/^(https|wss):/i.test(this.url)) return
        console.warn(
            `source-rpc: '${this.name}' is connecting to ${this.url} with allowInsecureTls, so the server's certificate is not checked. ` +
                'Anything able to answer on that address can read and rewrite this link. Use it for a development server, not a plant.'
        )
    }

    /**
     * A peer heard about through this link is routable through it - but only if nothing already
     * reaches it. A server that serves a peer locally must not start sending its traffic up to the
     * hub and back, and it would, since the hub lists that peer like any other.
     */
    private registerIfUnrouted(peer: string) {
        if (this.peerRegistry.get(peer)) return
        this.setKnownSource(peer)
    }

    private announce() {
        const announcement: PresenceAnnouncement = { name: this.name }
        if (this.carrying.length) announcement.carrying = this.carrying
        if (this.shape) announcement.shape = this.shape
        // So the server can address this peer before it has sent a frame - an event pushed to a
        // subscriber being the ordinary case. See PresenceAnnouncement.v.
        if (this.frameVersion !== 1) announcement.v = this.frameVersion
        this.socket?.emit(PRESENCE_EVENT, announcement)
    }

    /** The description hash this peer announces. See PresenceAnnouncement.shape. */
    private shape?: string

    /**
     * Set what this peer's surface hashes to, re-announcing if the change happens on a live link -
     * which it does whenever something is exposed after ready(), the way the introspection
     * namespace itself is.
     */
    announceShape(shape: string) {
        if (this.shape === shape) return
        this.shape = shape
        if (this.connected && this.announcePresence) this.announce()
    }

    /**
     * Say which peers can be reached through this connection. Sent again whenever the set changes,
     * which is how a peer appearing three hops away eventually becomes addressable from here.
     */
    advertise(peers: string[]) {
        const next = [...peers].sort()
        if (next.length === this.carrying.length && next.every((peer, index) => peer === this.carrying[index])) return
        this.carrying = next
        if (this.connected && this.announcePresence) this.announce()
    }

    /** The server's view of who else is connected, turned into the same events MQTT emits. */
    /** Checked and deduplicated before anyone hears about it - see TransportEvent.peerShape. */
    private noteShape(peer: string, shape: unknown) {
        if (!isUsableShape(shape)) return
        if (this.peerRegistry.noteShape(peer, shape)) this.emit(TransportEvent.peerShape, peer, shape)
    }

    private onPresence(update: PresenceUpdate) {
        if (Array.isArray(update.peers)) {
            // The snapshot is here, whoever asked for it. Resolved before the loop, not after:
            // even an all-filtered list (every name unusable or already known) is still the sweep.
            const landed = this.sweepLanded
            this.sweepLanded = undefined
            landed?.()
            for (const peer of update.peers) {
                if (!isUsablePeerName(peer) || peer === this.name) continue
                // Before the known-peers gate: a reconnect's fresh snapshot repeats known names,
                // and a repeated name carrying a new hash is the restart this exists to catch.
                if (update.shapes && Object.prototype.hasOwnProperty.call(update.shapes, peer)) this.noteShape(peer, update.shapes[peer])
                if (this.knownPeers.has(peer)) continue
                this.knownPeers.add(peer)
                this.registerIfUnrouted(peer)
                this.emit(TransportEvent.peerOnline, peer)
            }
            return
        }
        if (!isUsablePeerName(update.peer) || update.peer === this.name) return
        if (update.state === 'offline') {
            if (!this.knownPeers.delete(update.peer)) return
            this.emit(TransportEvent.peerGone, update.peer)
        } else {
            // Same reasoning as the snapshot: a re-announcement of a known peer with a new hash is
            // a change of surface, not a change of presence, and must not be gated out with it.
            if (update.shape !== undefined) this.noteShape(update.peer, update.shape)
            if (this.knownPeers.has(update.peer)) return
            this.knownPeers.add(update.peer)
            this.registerIfUnrouted(update.peer)
            this.emit(TransportEvent.peerOnline, update.peer)
        }
    }

    /**
     * Hand a frame to this peer's own handlers, or on to whichever transport reaches its addressee.
     * The second case is what makes a server that is both a hub for its own peers and a member of a
     * bus work: a call for one of its peers arrives down this link and has to be passed inwards,
     * not answered here. Without it the frame reached the right process and was refused by it.
     */
    private async deliver(message: Message, source: string, target: string, hops: number) {
        if (target !== this.name) {
            const carrier = this.peerRegistry.get(target)
            if (carrier && carrier !== (this as IGenericModule) && carrier.isTransport()) {
                if (hops + 1 > MAX_RELAY_HOPS) {
                    await refuseDelivery(this, message, source, target, 'TransportError', `over ${MAX_RELAY_HOPS} relays`)
                    return
                }
                const relay = carrier as { forward?: (message: Message, source: string, target: string, hops: number) => void }
                if (relay.forward) relay.forward(message, source, target, hops + 1)
                else await carrier.receive(message, source, target)
                return
            }
        }
        if (this.targetExists(target)) {
            await this.send(message, source, target)
            return
        }
        await refuseDelivery(this, message, source, target, 'TransportError', `no route to '${target}'`)
    }

    /** Send over this link with a hop count, for a frame being passed along rather than originated. */
    forward(message: Message, source: string, target: string, hops: number) {
        try {
            this.emitFrame(this.requireSocket(), message, source, target, hops)
        } catch (e) {
            // Relaying is done for someone else, so there is no caller here to reject.
            this.emit(TransportEvent.unroutable, { source, target, reason: `cannot forward: ${String(e)}`, error: e })
        }
    }

    /**
     * Put one frame on the link in whichever layout this peer sends.
     *
     * Throws rather than returning quietly when the message has no representation. The old path
     * could not fail this way - it encoded a `Message` whole and asked no questions - so a caller
     * whose message cannot be framed has to hear about it here, where there is still a call to
     * reject, rather than discover it as a timeout.
     */
    private emitFrame(socket: Socket, message: Message, source: string, target: string, hops = 0) {
        if (this.frameVersion === 1) {
            const header = this.buildHeader(source, target, hops ? { hops } : undefined)
            socket.emit(LEGACY_FRAME_EVENT, this.frameMessage(header, this.codec.encode(message)))
            return
        }
        const wire = toWireFrame(message, source, target, hops)
        if (!wire) throw new Error(`SocketIoClientTransport '${this.name}': no frame representation for this message`)
        socket.emit(FRAME_EVENT, this.codec.encode(wire))
    }

    /**
     * The link, or an error saying there is none - where "there is none" includes a socket that
     * exists but is not currently connected.
     *
     * Sending went through `this.socket?.emit(...)`, which is a no-op once the transport is closed -
     * so an outgoing call was discarded without a word and its caller waited out the full timeout
     * for a frame that was never going to be sent.
     *
     * The connected check is the second half of the same idea and was the more expensive omission.
     * socket.io buffers an emit made while disconnected and flushes the whole buffer on reconnect,
     * which reads like resilience and is not: by then the call has failed on its own timeout, its
     * caller has been told so and acted on it, and the frame is delivered anyway. Measured at nine
     * and a half seconds past the caller's deadline on a two-second timeout, with the command
     * running at the far end. The server's own deadline re-read cannot catch it either, because
     * that budget is measured from the moment a frame *arrives* and this one arrives untouched.
     */
    private requireSocket() {
        if (!this.socket) throw new Error(`SocketIoClientTransport '${this.name}': not connected to ${this.url ?? 'the default url'}`)
        if (!this.socket.connected) throw new Error(`SocketIoClientTransport '${this.name}': the link to ${this.url ?? 'the default url'} is down`)
        return this.socket
    }

    override async receive(message: Message, source: string, target: string) {
        // Refused while the link is down rather than waited out or handed over - see requireSocket.
        // The rule this replaces was "let socket.io buffer it, and if the link never comes back the
        // call fails on its own timeout", which is true and is not the whole story: when the link
        // *does* come back, the buffered frame is delivered to a caller that gave up long ago.
        this.emitFrame(this.requireSocket(), message, source, target)
    }

    override isTransport() {
        return true
    }
}
